BeardedSpice/Tabs/QuodLibetTabAdapter.m

```diff
@@ -0,0 +1,364 @@
+        return [NSString stringWithFormat:@"%02ld:%02ld:%02ld", (long)(min / 60), (long)(min % 60), (long)(ti % 60)];
+    }
+}
+
+
+@implementation QuodLibetTabAdapter{
+    QuodLibetStatus _status;
+    double _progress;
+    BSTrack *_track;
+}
+
+- (id)init {
+    self = [super init];
+    if (self) {
+        _status = QuodLibetStatusUnknown;
+        _progress = 0;
+        _track = nil;
+    }
+    return self;
+}
+
++ (NSString *)displayName{
+    return APPNAME_QUOD_LIBET;
+}
+
++ (NSString *)bundleId{
+    return APPID_QUOD_LIBET;
+}
+
+- (NSString *)URL{
+    return APPNAME_QUOD_LIBET;
+}
+
+- (NSString *)key{
+    return @"A:" APPNAME_QUOD_LIBET;
+}
+
+- (NSString *)title{
+    @autoreleasepool {
+        [self loadQuodLibetState];
+
+        NSString *title;
+        if (_status == QuodLibetStatusPlaying) {
+
+            BSTrack *trackInfo = [self trackInfo];
+            if (trackInfo.track)
+                title = trackInfo.track;
+
+            if (trackInfo.artist) {
+
+                if (title) title = [title stringByAppendingFormat:@" - %@", trackInfo.artist];
+                else
+                    title = trackInfo.artist;
+            }
+        }
+
+        if ([NSString isNullOrEmpty:title]) {
+            title = NSLocalizedString(@"No Track", @"QuodLibetTabAdapter");
+        }
+
+        return [NSString stringWithFormat:@"%@ (%@)", title, APPNAME_QUOD_LIBET];
+    }
+}
+
+#pragma mark Player control methods
+
+- (void)toggle{
+    [self sendControlToQuodLibet:QL_CONTROL_PLAY_PAUSE];
+}
+
+- (void)pause{
+    [self sendControlToQuodLibet:QL_CONTROL_PAUSE];
+}
+
+- (void)next{
+    [self sendControlToQuodLibet:QL_CONTROL_NEXT];
+}
+
+- (void)previous{
+    [self sendControlToQuodLibet:QL_CONTROL_PREVIOUS];
+}
+
+- (BSTrack *)trackInfo{
+    [self loadQuodLibetState];
+    if (_status == QuodLibetStatusPlaying) {
+        if (_track) {
+            return _track;
+        }
+        
+        // read track status from current file
+        NSError *error = nil;
+        NSString *current = [[NSString alloc] initWithContentsOfFile:[QL_CURRENT stringByExpandingTildeInPath]
+                                              encoding:NSUTF8StringEncoding error:&error];
+        if (error) {
+            NSLog(@"Failed to read QuodLibet current track info: %@", [error localizedDescription]);
+            return nil;
+        }
+        
+        _track = [BSTrack new];
+
+        NSArray *lines = [current componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
+        for (NSString *line in lines) {
+            NSRange delim = [line rangeOfString:@"="];
+            if (delim.location != NSNotFound) {
+                NSString *key = [line substringToIndex:delim.location];
+                NSString *value = [line substringFromIndex:delim.location + 1];
+                
+                if ([key isEqualToString:QL_TRACK_TITLE]) {
+                    _track.track = value;
```

Comment:
Everything's OK! But there are tracks with no metadata. What if to make the file name without the extension as the title of the track in the case when the track has no title.

Reply:
Fair point; Quod Libet, in that case, uses the filename too with `[Unknown]` added. I'll do the same here.